### PR TITLE
Fix NPE when setting a player's spectator target

### LINKED ISSUE #1595: NPE when setting a player's spectator target
### What behaviour is expected:
No NPE should be thrown.

### What behaviour is observed:
An NPE is thrown.

### Steps/models to reproduce:
Set the `spectatorTarget` of a player.

### Paper build number:
376+ (after spectatorTarget events were added.)

### Anything else:
I have found why this is a problem and will be submitting a pull request quickly.

### LINKED ISSUE #1603: Spectating does not work any more
### What behaviour is expected:
Spectating an entity should start when left clicking the entity in spectator mode

### What behaviour is observed:
Not spectating the entity but an exception in the server log

### Steps/models to reproduce:
- /gamemode spectator
- click an entity

### Plugin list:
No plugin needed

### Paper build number:
 git-Paper-382 (MC: 1.13.1)

### Anything else:
`[15:56:03 FATAL]: Error executing task
java.lang.NullPointerException: null
        at net.minecraft.server.v1_13_R2.PacketPlayOutCamera.<init>(SourceFile:19) ~[patched_1.13.1.jar:git-Paper-382]
        at net.minecraft.server.v1_13_R2.EntityPlayer.setSpectatorTarget(EntityPlayer.java:1398) ~[patched_1.13.1.jar:git-Paper-382]
        at net.minecraft.server.v1_13_R2.EntityPlayer.attack(EntityPlayer.java:1414) ~[patched_1.13.1.jar:git-Paper-382]
        at net.minecraft.server.v1_13_R2.PlayerConnection.a(PlayerConnection.java:1979) ~[patched_1.13.1.jar:git-Paper-382]
        at net.minecraft.server.v1_13_R2.PacketPlayInUseEntity.a(PacketPlayInUseEntity.java:49) ~[patched_1.13.1.jar:git-Paper-382]
        at net.minecraft.server.v1_13_R2.PacketPlayInUseEntity.a(PacketPlayInUseEntity.java:6) ~[patched_1.13.1.jar:git-Paper-382]
        at net.minecraft.server.v1_13_R2.PlayerConnectionUtils.lambda$ensureMainThread$0(PlayerConnectionUtils.java:12) ~[patched_1.13.1.jar:git-Paper-382]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[?:1.8.0_181]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[?:1.8.0_181]
        at net.minecraft.server.v1_13_R2.SystemUtils.a(SystemUtils.java:109) ~[patched_1.13.1.jar:git-Paper-382]
        at net.minecraft.server.v1_13_R2.MinecraftServer.b(MinecraftServer.java:1016) ~[patched_1.13.1.jar:git-Paper-382]
        at net.minecraft.server.v1_13_R2.DedicatedServer.b(DedicatedServer.java:439) ~[patched_1.13.1.jar:git-Paper-382]
        at net.minecraft.server.v1_13_R2.MinecraftServer.a(MinecraftServer.java:943) ~[patched_1.13.1.jar:git-Paper-382]
        at net.minecraft.server.v1_13_R2.MinecraftServer.run(MinecraftServer.java:841) ~[patched_1.13.1.jar:git-Paper-382]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_181]
`

### PR DESCRIPTION
**Pull request by CalebBassham**
Fix NPE from being thrown when setting a player's spectator target. It was occurring because in #1498 the method was changed to add some events. Basically, I was trying to send the `PacketPlayOutCamera` packet to the player using `this.cv` which is the current spectator target but it had not been set yet; therefore, the NPE was thrown. I have updated the method to set the field before sending the packet. 

Closes #1595 
Closes #1603

### CODE REVIEW

**Comment by electronicboy** [comment]
Your PR has various issues in terms of the patch having changes to imports as well as formatting changes in stuff irrelevant to this patch

File: Spigot-Server-Patches/0402-Fix-NPE-when-setting-a-player-s-spectator-target.patch [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+@@ -62,9 +54,19 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
+     private EntityHuman.EnumChatVisibility cs;
+     private boolean ct = true;
+     private long cu = SystemUtils.b();
+-    private Entity cv; private void setSpectatorTargetField(Entity e) { this.cv = e; } // Paper - OBFHELPER
++    private Entity cv;
++
++    private void setSpectatorTargetField(Entity e) {
++        this.cv = e;
++    } // Paper - OBFHELPER
```
Comment (hugmanrique) [inline Spigot-Server-Patches/0402-Fix-NPE-when-setting-a-player-s-spectator-target.patch]: Avoid formatting obfuscation helpers, they should be in the same line as the field the OBFHelper is going to modify.

File: Spigot-Server-Patches/0402-Fix-NPE-when-setting-a-player-s-spectator-target.patch [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
++
+     public boolean worldChangeInvuln;
+-    private boolean cx; private void setHasSeenCredits(boolean has) { this.cx = has; } // Paper - OBFHELPER
++    private boolean cx;
++
++    private void setHasSeenCredits(boolean has) {
```
Comment (hugmanrique) [inline Spigot-Server-Patches/0402-Fix-NPE-when-setting-a-player-s-spectator-target.patch]: Same thing here

File: Spigot-Server-Patches/0402-Fix-NPE-when-setting-a-player-s-spectator-target.patch [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+     private Vec3D cz;
+     private int cA;
+@@ -535,8 +537,8 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
+             this.playerConnection.a((Packet) (new PacketPlayOutCombatEvent(this.getCombatTracker(), PacketPlayOutCombatEvent.EnumCombatEventType.ENTITY_DIED, ichatbasecomponent)), (future) -> {
+                 if (!future.isSuccess()) {
+                     String s = ichatbasecomponent.a(256);
+-                    ChatMessage chatmessage = new ChatMessage("death.attack.message_too_long", new Object[] { (new ChatComponentText(s)).a(EnumChatFormat.YELLOW)});
+-                    IChatBaseComponent ichatbasecomponent1 = (new ChatMessage("death.attack.even_more_magic", new Object[] { this.getScoreboardDisplayName()})).a((chatmodifier) -> {
++                    ChatMessage chatmessage = new ChatMessage("death.attack.message_too_long", new Object[]{(new ChatComponentText(s)).a(EnumChatFormat.YELLOW)});
++                    IChatBaseComponent ichatbasecomponent1 = (new ChatMessage("death.attack.even_more_magic", new Object[]{this.getScoreboardDisplayName()})).a((chatmodifier) -> {
```
Comment (hugmanrique) [inline Spigot-Server-Patches/0402-Fix-NPE-when-setting-a-player-s-spectator-target.patch]: Avoid modifying the formatting of code this patch isn't touching

File: Spigot-Server-Patches/0402-Fix-NPE-when-setting-a-player-s-spectator-target.patch [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+                         chatmodifier.setChatHoverable(new ChatHoverable(ChatHoverable.EnumHoverAction.SHOW_TEXT, chatmessage));
+                     });
+ 
+@@ -687,7 +689,8 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
+             this.world.kill(this);
+             if (!this.viewingCredits) {
+                 this.viewingCredits = true;
+-                if (world.paperConfig.disableEndCredits) this.setHasSeenCredits(true); // Paper - Toggle to always disable end credits
++                if (world.paperConfig.disableEndCredits)
++                    this.setHasSeenCredits(true); // Paper - Toggle to always disable end credits
```
Comment (hugmanrique) [inline Spigot-Server-Patches/0402-Fix-NPE-when-setting-a-player-s-spectator-target.patch]: Same thing here

File: Spigot-Server-Patches/0402-Fix-NPE-when-setting-a-player-s-spectator-target.patch [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+                 this.playerConnection.sendPacket(new PacketPlayOutGameStateChange(4, this.cx ? 0.0F : 1.0F));
+                 this.cx = true;
+             }
+@@ -798,7 +801,8 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
+         return super.isInvulnerable(damagesource) || this.H();
+     }
+ 
+-    protected void a(double d0, boolean flag, IBlockData iblockdata, BlockPosition blockposition) {}
++    protected void a(double d0, boolean flag, IBlockData iblockdata, BlockPosition blockposition) {
```
Comment (hugmanrique) [inline Spigot-Server-Patches/0402-Fix-NPE-when-setting-a-player-s-spectator-target.patch]: Here

File: Spigot-Server-Patches/0402-Fix-NPE-when-setting-a-player-s-spectator-target.patch (hunk 1 above)
Comment (CalebBassham) [reply Spigot-Server-Patches/0402-Fix-NPE-when-setting-a-player-s-spectator-target.patch]: Yea I am already fixing it right now sorry. IntelliJ formatted it and I didn't realize.

File: Spigot-Server-Patches/0402-Fix-NPE-when-setting-a-player-s-spectator-target.patch [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
++    }
+ 
+     protected void b(BlockPosition blockposition) {
+         if (!this.isSpectator()) {
+@@ -843,7 +847,7 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
+         if (false && itileentitycontainer instanceof ILootable && ((ILootable) itileentitycontainer).getLootTable() != null && this.isSpectator()) {
+             this.a((new ChatMessage("container.spectatorCantOpen", new Object[0])).a(EnumChatFormat.RED), true);
+         } else {
+-            boolean cancelled = itileentitycontainer instanceof ILootable && ((ILootable) itileentitycontainer).getLootTable()!= null && this.isSpectator();
```
Comment (hugmanrique) [inline Spigot-Server-Patches/0402-Fix-NPE-when-setting-a-player-s-spectator-target.patch]: Here (there are more unneeded changes below, not gonna comment on those)

**Review by hugmanrique — CHANGES_REQUESTED** [review]

File: Spigot-Server-Patches/0402-Fix-NPE-when-setting-a-player-s-spectator-target.patch (hunk 2 above)
Comment (CalebBassham) [reply Spigot-Server-Patches/0402-Fix-NPE-when-setting-a-player-s-spectator-target.patch]: Yea I am already fixing it right now sorry. IntelliJ formatted it and I didn't realize.

File: Spigot-Server-Patches/0402-Fix-NPE-when-setting-a-player-s-spectator-target.patch (hunk 3 above)
Comment (CalebBassham) [reply Spigot-Server-Patches/0402-Fix-NPE-when-setting-a-player-s-spectator-target.patch]: Yea I am already fixing it right now sorry. IntelliJ formatted it and I didn't realize.

File: Spigot-Server-Patches/0402-Fix-NPE-when-setting-a-player-s-spectator-target.patch (hunk 4 above)
Comment (CalebBassham) [reply Spigot-Server-Patches/0402-Fix-NPE-when-setting-a-player-s-spectator-target.patch]: Yea I am already fixing it right now sorry. IntelliJ formatted it and I didn't realize.

File: Spigot-Server-Patches/0402-Fix-NPE-when-setting-a-player-s-spectator-target.patch (hunk 5 above)
Comment (CalebBassham) [reply Spigot-Server-Patches/0402-Fix-NPE-when-setting-a-player-s-spectator-target.patch]: Yea I am already fixing it right now sorry. IntelliJ formatted it and I didn't realize.

File: Spigot-Server-Patches/0402-Fix-NPE-when-setting-a-player-s-spectator-target.patch (hunk 6 above)
Comment (CalebBassham) [reply Spigot-Server-Patches/0402-Fix-NPE-when-setting-a-player-s-spectator-target.patch]: Yea I am already fixing it right now sorry. IntelliJ formatted it and I didn't realize.

File: Spigot-Server-Patches/0402-Fix-NPE-when-setting-a-player-s-spectator-target.patch [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+-            this.playerConnection.a(this.cv.locX, this.cv.locY, this.cv.locZ, this.yaw, this.pitch, TeleportCause.SPECTATE); // CraftBukkit
++            if (!playerStartSpectatingEntityEvent.callEvent()) {
++                return;
++            }
+         }
+ 
+         setSpectatorTargetField(newSpectatorTarget);
++
++        this.playerConnection.sendPacket(new PacketPlayOutCamera(newSpectatorTarget));
++        this.playerConnection.a(this.cv.locX, this.cv.locY, this.cv.locZ, this.yaw, this.pitch, TeleportCause.SPECTATE); // CraftBukkit
```
Comment (BillyGalbreath) [inline Spigot-Server-Patches/0402-Fix-NPE-when-setting-a-player-s-spectator-target.patch]: Forgot to mark with `// Paper start` and `// Paper end` your actual changes

**Comment by electronicboy** [comment]
This pr appears to be fixing your previous pr from #1498 and should be made in that patch, the easiest way to do this at this point is to amend your current patch and then `./paper e server` and then move your new patch to below the older one and replace `edit` with `f` for fixup, and then save and quit the editor and rebuild the patches once it's finished

if you need any assistance with this, feel free to reach out to us on IRC/Discord, links are on the readme file

**Comment by CalebBassham** [comment]
I think that's what you wanted.

File: Spigot-Server-Patches/0397-Call-player-spectator-target-events.patch [hunk 8]
```diff
@@ -27,30 +27,33 @@ index 7ac07ac07ac0..7ac07ac07ac0 100644
...
++        }
 +
-             this.playerConnection.sendPacket(new PacketPlayOutCamera(this.cv));
-             this.playerConnection.a(this.cv.locX, this.cv.locY, this.cv.locZ, this.yaw, this.pitch, TeleportCause.SPECTATE); // CraftBukkit
-         }
- 
 +        setSpectatorTargetField(newSpectatorTarget);
++
++        this.playerConnection.sendPacket(new PacketPlayOutCamera(newSpectatorTarget));
++        this.playerConnection.a(this.cv.locX, this.cv.locY, this.cv.locZ, this.yaw, this.pitch, TeleportCause.SPECTATE); // CraftBukkit
```
Comment (BillyGalbreath) [inline Spigot-Server-Patches/0397-Call-player-spectator-target-events.patch]: Still no `// Paper` tags on your actual changes

Comment (CalebBassham) [reply Spigot-Server-Patches/0397-Call-player-spectator-target-events.patch]: Adding it now

**Comment by CalebBassham** [comment]
Sorry for all of the problems this has caused on a simple change. Hopefully I got everything correct now!

**Comment by hugmanrique** [comment]
Don't worry about it, we all make mistakes! You wouldn't believe how many of them I made on one of my first contributions to Paper, but you learn from them 😄

**Pull request merged**